android/src/main/java/com/tailscale/ipn/ui/notifier/Notifier.kt

```diff
@@ -26,116 +26,72 @@ class Watcher(
-    private external fun stopIPNBusWatcher(sessionId: String)
-
-    private var watchers = HashMap<String, Watcher>()
-
-    // Callback from jni when a new notification is received
-    fun onNotify(notification: String, sessionId: String) {
-        val notify = decoder.decodeFromString<Notify>(notification)
-        val watcher = watchers[sessionId]
-        watcher?.let { watcher.callback(notify) } ?: {
-            Log.e(
-                "Notifier",
-                "Received notification for unknown session: ${sessionId}"
-            )
-        }
+object Notifier {
+    private val TAG = Notifier::class.simpleName
+    private val decoder = Json { ignoreUnknownKeys = true }
+    private val isReady = CompletableDeferred<Boolean>()
+
+    val state: StateFlow<Ipn.State> = MutableStateFlow(Ipn.State.NoState)
+    val netmap: StateFlow<Netmap.NetworkMap?> = MutableStateFlow(null)
+    val prefs: StateFlow<Ipn.Prefs?> = MutableStateFlow(null)
+    val engineStatus: StateFlow<Ipn.EngineStatus?> = MutableStateFlow(null)
+    val tailFSShares: StateFlow<Map<String, String>?> = MutableStateFlow(null)
+    val browseToURL: StateFlow<String?> = MutableStateFlow(null)
+    val loginFinished: StateFlow<String?> = MutableStateFlow(null)
+    val version: StateFlow<String?> = MutableStateFlow(null)
+
+    // Called by the backend when the localAPI is ready to accept requests.
+    @JvmStatic
+    @Suppress("unused")
+    fun onReady() {
+        isReady.complete(true)
+        Log.d(TAG, "Ready")
     }
 
-    // Watch the IPN bus for notifications
-    // Notifications will be passed to the caller via the callback until
-    // the caller calls unwatchIPNBus with the sessionId returned from this call.
-    private fun watchIPNBus(mask: Int, callback: NotifierCallback): String {
-        val sessionId = generateSessionId()
-        val watcher = Watcher(sessionId, mask, callback)
-        watchers[sessionId] = watcher
+    fun start(scope: CoroutineScope) {
+        Log.d(TAG, "Starting")
         scope.launch(Dispatchers.IO) {
             // Wait for the notifier to be ready
             isReady.await()
-            Log.d("Notifier", "Starting IPN Bus watcher for sessionid: ${sessionId}")
-            startIPNBusWatcher(sessionId, mask)
-            watchers.remove(sessionId)
-            Log.d("Notifier", "IPN Bus watcher for sessionid:${sessionId} has halted")
-        }
-        return sessionId
-    }
-
-    // Cancels the watcher with the given sessionId. No errors are thrown or
-    // indicated for invalid sessionIds.
-    private fun unwatchIPNBus(sessionId: String) {
-        stopIPNBusWatcher(sessionId)
-    }
-
-    // Cancels all watchers
-    fun cancelAllWatchers() {
-        for (sessionId in watchers.values.map({ it.sessionId })) {
-            unwatchIPNBus(sessionId)
+            val mask =
+                NotifyWatchOpt.Netmap.value or NotifyWatchOpt.Prefs.value or NotifyWatchOpt.InitialState.value
```

Comment:
Nit: A comment here explaining the mask choice might be useful, but it should be easy enough for the next guy to figure out.